compiler/src/iree/compiler/Codegen/LLVMCPU/test/tile-root-fuse-consumer-producer.mlir

```diff
@@ -0,0 +1,151 @@
+// RUN: iree-opt --pass-pipeline="builtin.module(func.func(iree-llvmcpu-tile-root-and-fuse-producer-consumer{tiling-level=0}), canonicalize)"  --split-input-file %s | FileCheck %s
+
+#config1 = #iree_codegen.lowering_config<tile_sizes = [[1, 0, 0, 0, 0, 0], [1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0], [1, 0, 0, 16, 16, 0], [0, 0, 1, 0, 0, 1], [0, 0, 0, 0, 0, 0]]>
+#map = affine_map<(d0, d1, d2, d3) -> (d0, d1, d2, d3)>
+#map1 = affine_map<(d0, d1, d2, d3) -> (d0, d2)>
+func.func @mmt4d_bias_relu(%arg0: tensor<?x?x16x1xf32>, %arg1: tensor<?x?x16x1xf32>, %arg2: tensor<?x16xf32>) -> tensor<?x?x16x16xf32> {
+  %cst = arith.constant 0.000000e+00 : f32
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %dim = tensor.dim %arg0, %c0 : tensor<?x?x16x1xf32>
+  %dim_0 = tensor.dim %arg1, %c1 : tensor<?x?x16x1xf32>
+  %0 = tensor.empty(%dim, %dim_0) : tensor<?x?x16x16xf32>
+  %1 = tensor.empty(%dim, %dim_0) : tensor<?x?x16x16xf32>
+  %2 = linalg.fill ins(%cst : f32) outs(%1 : tensor<?x?x16x16xf32>) -> tensor<?x?x16x16xf32>
+  %3 = linalg.mmt4d {lowering_config = #config1} ins(%arg0, %arg1 : tensor<?x?x16x1xf32>, tensor<?x?x16x1xf32>) outs(%2 : tensor<?x?x16x16xf32>) -> tensor<?x?x16x16xf32>
+  %4 = linalg.generic {indexing_maps = [#map, #map1, #map], iterator_types = ["parallel", "parallel", "parallel", "parallel"]} ins(%3, %arg2 : tensor<?x?x16x16xf32>, tensor<?x16xf32>) outs(%1 : tensor<?x?x16x16xf32>) {
+  ^bb0(%in: f32, %in_1: f32, %out: f32):
+    %5 = arith.addf %in, %in_1 : f32
+    %6 = arith.maximumf %5, %cst : f32
+    linalg.yield %6 : f32
+  } -> tensor<?x?x16x16xf32>
+  return %4 : tensor<?x?x16x16xf32>
+}
+//      CHECK: func.func @mmt4d_bias_relu(
+//      CHECK:   scf.for
+// CHECK-SAME:   {
+//      CHECK:     linalg.fill
+//      CHECK:     linalg.mmt4d
+//      CHECK:     linalg.generic
+//      CHECK:   }
+
+// -----
+
+/// This example is used to test the lowering of the diamond pattern.
+///           -> generic (add) --- \
+/// matmul                             generic(add)
+
+///           -> generic (max) --- /
+/// TODO: This is not supported yet.
```

Comment:
Can you explain why it is not supported yet (IIUC the tiling interface API does not support multi-use insert_slice ops). It will make it easier for someone to fix the test after LLVM integration if there is a clear explanation for what the ideal test output is and why it does not work right now.

Reply:
Where is it coming from? I'm worried that we are solving problems that are not expected. I can't image the case that we form the dispatch in this way.

The only case I can think of for today is softmax. Because they are decomposed to few generic ops, see below. Today the second reduction is the root op (which is not a hard limit IMO). I wonder if the pass works for the case or not. ([Here](https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Codegen/Common/test/decompose_softmax.mlir) is the lit test for softmax decomposition.)

(side note: IMO, we should defer the softmax decomposition to either before reduction tiling or vectorization. We do have tiling implementation for softmax, so I don't really see a reason to decompose it before codegen strategy selection pass. It is not part of the scope of the work, so we don't need to make it happen now. It is just my opinion about how we can codegen softmax in CPU pipelines.)

```
reduction ----> reduction
 |                 |
 |                 v
 ------------> elementwise
```

Reply:
I would rather rely on elementwise operations to fuse the `add/max/add` operations above and have a single operation consumer to deal with.

Reply:
I'll remove the test case.